Add join condition & inner/outer join support to NestedLoopJoin

usurai opened the pull request:
1. Update PlanBuilder to build nested loop join with join condition.
2. Update NestedLoopJoinProbe to support join condition and inner / outer join types.
3. Update test and documentation.

mbasmanova commented:
@usurai Thank you for working on this.

> This PR also fixes a bug in CrossJoinProbe: https://github.com/facebookincubator/velox/issues/4600

Would it be possible to extract this fix into its own PR?

mbasmanova commented:
> using a filter as join condition is simple, but also makes it hard to enforcing rules on join condition.

You can compile the expression into an ExprSet, then check whether the expression propagates nulls using Expr::propagatesNulls() API.

You can also check which fields the expression depends on using Expr::distinctFields() API and assert that there is at least one field from left and one field from right.

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1524,6 +1524,54 @@ class CrossJoinNode : public PlanNode {
+  NestedLoopJoinNode(
+      const PlanNodeId& id,
+      JoinType joinType,
+      TypedExprPtr joinCondition,
```
Let's remove CrossJoinNode and use NestedLoopJoinNode with nullptr for 'joinCondition'.

usurai replied:
Have removed as suggested.
1. For compatibility, I kept interfaces that were used to create `CrossJoinNode`, only updated them to create null-joinCondition `NestedLoopJoinNode`:
- `PlanBuilder::crossJoin`
- `PlanNodePtr toVeloxPlan(::duckdb::LogicalCrossProduct& logicalCrossProduct, ...)`

2. As `CrossJoinProbe` is the subset of `NestedLoopJoinProbe` in terms of functionality, maybe we can remove it as well?

mbasmanova replied:
@usurai Thank you for the follow-up. Let's remove references to CrossJoinNode/Operator completely.

usurai replied:
@mbasmanova I have replaced `CrossJoinProbe` with `NestedLoopJoinProbe` and renamed `CrossJoin{Bridge/Build}` to `NestedLoopJoin{Bridge/Build}`.

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1524,6 +1524,54 @@ class CrossJoinNode : public PlanNode {
+
+ private:
+  void addDetails(std::stringstream& stream) const override {
+    stream << joinTypeName(joinType_)
```
Let's move this implementation to the .cpp file.

mbasmanova commented on velox/exec/CrossJoinBuild.h:
```diff
@@ -37,6 +37,11 @@ class CrossJoinBuild : public Operator {
+  CrossJoinBuild(
+      int32_t operatorId,
+      DriverCtx* driverCtx,
+      std::shared_ptr<const core::NestedLoopJoinNode> joinNode);
```
pass by reference

mbasmanova reviewed (commented):
@usurai Some initial comments.

usurai commented:
@mbasmanova Thanks for reviewing.

> > This PR also fixes a bug in CrossJoinProbe: #4600
> 
> Would it be possible to extract this fix into its own PR?

I created a separate PR #4625 




> 
> You can compile the expression into an ExprSet, then check whether the expression propagates nulls using Expr::propagatesNulls() API.

I've thought about this, but didn't do that because this can only cover join conditions that join single key. For multi-key joins, null of certain key will cause the expression eval to false, but not null.

> 
> You can also check which fields the expression depends on using Expr::distinctFields() API and assert that there is at least one field from left and one field from right.

I think we can use expression like `u0 > 1 AND t0 < 10` to bypass the check.

mbasmanova commented:
>  using a filter as join condition is simple, but also makes it hard to enforcing rules on join condition.

@usurai I get your points. These makes sense. At a high-level, Velox is not expected to optimize the query plan, rather the application / caller is supposed to produce fully optimize plan before passing it Velox. Hence, it is Ok.

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1422,14 +1422,27 @@ class MergeJoinNode : public AbstractJoinNode {
-class CrossJoinNode : public PlanNode {
+/// Represents inner/outer nested loop joins. Translates to an
+/// exec::NestedLoopJoinProbe and exec::NestedLoopJoinBuild. A separate pipeline
+/// is produced for the build side when genreating exec::Operators. Nested loop
```
typo: genreating

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1422,14 +1422,27 @@ class MergeJoinNode : public AbstractJoinNode {
+/// Represents inner/outer nested loop joins. Translates to an
+/// exec::NestedLoopJoinProbe and exec::NestedLoopJoinBuild. A separate pipeline
+/// is produced for the build side when genreating exec::Operators. Nested loop
+/// join supports both euqal and non-equal joins. Expressions specified in
```
typo: euqal

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -1422,14 +1422,27 @@ class MergeJoinNode : public AbstractJoinNode {
+      TypedExprPtr joinCondition,
+      PlanNodePtr left,
+      PlanNodePtr right,
+      const RowTypePtr& outputType);
```
pass by value and std::move

usurai replied:
I find inconsistent usage of `outputType` parameter across `PlanNode.h`:
- const reference is used by `ArrowStreamNode`, `TableScanNode` and `TableWriteNode`
- left value is used by `PartitionedOutputNode`
- const value is used by `AbstractJoinNode` and hence `HashJoinNode`/`MergeJoinNode`
Should we unify them all to pass by value? (take ownership)

mbasmanova replied:
It would be nice to unify these.

usurai replied:
Got it, will do.

usurai replied:
Here is the PR #4683.

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -846,35 +846,90 @@ PlanNodePtr MergeJoinNode::create(const folly::dynamic& obj, void* context) {
+  VELOX_CHECK(
+      core::isInnerJoin(joinType_) || core::isLeftJoin(joinType_) ||
+          core::isRightJoin(joinType_) || core::isFullJoin(joinType_),
+      "Currently NestedLoopJoin only supports inner and outer join");
```
drop "Currently" and include 'joinType' in the error message for easier troubleshooting:

```
NestedLoopJoin only supports inner, left, right and full joins: {}
```

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -846,35 +846,90 @@ PlanNodePtr MergeJoinNode::create(const folly::dynamic& obj, void* context) {
+  TypedExprPtr joinCondition;
+  if (obj.count("joinCondition")) {
+    joinCondition =
+        ISerializable::deserialize<ITypedExpr>(obj["joinCondition"]);
```
Deserialization of ITypedExpr may need access to memory pool, hence, need to pass 'context'.

usurai replied:
Several invocations of `deserialize<ITypedExpr>` in `PlanNode.cpp` do not pass `context`, shoud we fix them?

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L802

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L833

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L1057

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L1062

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L1575

mbasmanova replied:
Yes, please.

mbasmanova commented on velox/exec/tests/PlanNodeSerdeTest.cpp:
```diff
@@ -100,6 +100,35 @@ TEST_F(PlanNodeSerdeTest, crossJoin) {
   testSerde(plan);
 }
 
+TEST_F(PlanNodeSerdeTest, nestedLoopJoin) {
```
Should we also update TEST_F(PlanNodeSerdeTest, crossJoin) ?

I expect PlanBuilder::crossJoin to be renamed to nestedLoopJoin.

mbasmanova commented on velox/exec/tests/PlanNodeSerdeTest.cpp:
```diff
@@ -100,6 +100,35 @@ TEST_F(PlanNodeSerdeTest, crossJoin) {
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .values({left})
+          .nestedLoopJoin(
```
Would you add a test with non-default join type to make sure it is preserved during serde?

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -846,35 +846,90 @@ PlanNodePtr MergeJoinNode::create(const folly::dynamic& obj, void* context) {
       sources_({std::move(left), std::move(right)}),
-      outputType_(std::move(outputType)) {}
+      outputType_(outputType) {
+  VELOX_CHECK(
```
These should be user errors, hence, VELOX_USER_XXX

mbasmanova commented on velox/exec/tests/PlanNodeToStringTest.cpp:
```diff
@@ -389,9 +389,9 @@ TEST_F(PlanNodeToStringTest, crossJoin) {
+  ASSERT_EQ("-- NestedLoopJoin\n", plan->toString());
   ASSERT_EQ(
-      "-- CrossJoin[] -> t_c0:SMALLINT, t_c1:INTEGER, u_c1:INTEGER\n",
+      "-- NestedLoopJoin[INNER] -> t_c0:SMALLINT, t_c1:INTEGER, u_c1:INTEGER\n",
```
Let's also add a test case for when there is join condition.

mbasmanova reviewed (commented):
@usurai PlanNode changes look good. This PR is very large. Would it be possible to extract PlanNode changes into a separate PR? In that PR, the translation can use VELOX_NYI if joinCondition is specified. This would allow us to review and merge smaller PRs.

mbasmanova commented on velox/exec/Task.h:
```diff
@@ -29,7 +29,7 @@ namespace facebook::velox::exec {
 
 class HashJoinBridge;
-class CrossJoinBridge;
+class NestedLoopJoinBridge;
```
To help further reduce the size of this PR, consider extracting renaming of CrossJoinProbe/Build/Bridge into a separate PR.

usurai replied:
#4678 is for renaming, thanks.

usurai commented:
> @usurai PlanNode changes look good. This PR is very large. Would it be possible to extract PlanNode changes into a separate PR? In that PR, the translation can use VELOX_NYI if joinCondition is specified. This would allow us to review and merge smaller PRs.

Hi @mbasmanova, sorry for the large PR. #4665 is the extraction for PlanNode part changes, I've addressed comments in that PR and mark resolved here. Thanks!

usurai commented:
Hi @mbasmanova, since nested loop join implementation and its test uses some functions/state from HashProbe/HashJoinTest, I created a separate PR #4693 to extract them, please take a look. Thanks.

usurai commented:
@mbasmanova @xiaoxmeng I added join condition support in the latest commit of this PR, please take a look.

xiaoxmeng commented on velox/exec/tests/utils/PlanBuilder.h:
```diff
@@ -619,6 +619,23 @@ class PlanBuilder {
       const std::vector<std::string>& outputLayout,
       core::JoinType joinType = core::JoinType::kInner);
 
+  /// Add a NestedLoopJoinNode to join two inputs using filter as join
```
```
s/Add/Adds/
```

usurai replied:
34 out of 36 of function comments in PlanBuilder use 'Add', maybe open a new PR to change them all?

xiaoxmeng commented on velox/exec/OperatorUtils.h:
```diff
@@ -136,4 +136,13 @@ folly::Range<vector_size_t*> initializeRowNumberMapping(
     vector_size_t size,
     memory::MemoryPool* pool);
 
+// Projects children of 'src' RowVector to 'dest' RowVector according to
```
```
nit: s/RowVector/row vector/
```

xiaoxmeng replied:
Can you add unit test for this? Thanks!

xiaoxmeng commented on velox/exec/OperatorUtils.h:
```diff
@@ -136,4 +136,13 @@ folly::Range<vector_size_t*> initializeRowNumberMapping(
+    RowVectorPtr dest,
+    RowVectorPtr src,
+    const std::vector<IdentityProjection>& projections,
+    int32_t size,
```
What about size? Could you comment? Thanks!

xiaoxmeng commented on velox/exec/OperatorUtils.cpp:
```diff
@@ -396,4 +396,15 @@ folly::Range<vector_size_t*> initializeRowNumberMapping(
+    const std::vector<IdentityProjection>& projections,
+    int32_t size,
+    const BufferPtr mapping) {
+  for (const auto projection : projections) {
```
```
s/const auto/const auto&/
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,120 @@ class NestedLoopJoinProbe : public Operator {
   /// Maximum number of rows in the output batch.
   const uint32_t outputBatchSize_;
+  const core::JoinType joinType_;
+  ProbeOperatorState state_{ProbeOperatorState::kWaitForBuild};
```
nit: leave an empty line between const and non-const? Thanks!

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,120 @@ class NestedLoopJoinProbe : public Operator {
+  // buildIndex_ updates.
+  vector_size_t prevProbeCnt_{0};
+  bool lastProbe_{false};
+  // Represents whether probe side row has been matched
```
```
s/row has/rows have/
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,120 @@ class NestedLoopJoinProbe : public Operator {
+  size_t buildIndex_{0};
+  std::vector<IdentityProjection> buildProjections_;
+  BufferPtr buildIndices_;
+  // Represents whether probe build row has been matched
```
Leave an empty line between different set of fields? thanks!

xiaoxmeng replied:
```
s/probe build row/probed build rows/
s/matched/matched./
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,67 +54,290 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+}
+
+bool NestedLoopJoinProbe::getBuildData(ContinueFuture* future) {
+  VELOX_CHECK(buildData_.has_value() == false);
```
```
VELOX_CHECK(!buildData_.has_value());
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +346,134 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+  return output;
 }
 
-void NestedLoopJoinProbe::close() {
```
Why remove this close method? Could you add to clear all the major objects that might use a lot memory here? We might abort operator early. Thanks!

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,67 +54,290 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+RowVectorPtr NestedLoopJoinProbe::getOutput() {
+  while (true) {
+    switch (state_) {
+      case ProbeOperatorState::kFinish: {
```
```
case ProbeOperatorState::kFinish:
  return nullptr;
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,120 @@ class NestedLoopJoinProbe : public Operator {
+  // Return true if buildIndex_ is points to the end of buildData_.
+  bool advanceIndex(vector_size_t probeCnt);
+
+  bool buildIndexAtEnd() const {
```
```
nit: s/buildIndexAtEnd/hasProbedAllBuildData/
```

usurai replied:
Changed.
Note that for full/right joins, after probing, buildIndex_ is reset and reused for for emitting mis-matched build side data. This method is also used in this scenario.

xiaoxmeng reviewed (commented):
@usurai thanks for the change. Left some comments and will continue to review.

xiaoxmeng reviewed (commented):
@usurai thanks for the change. Left some comments and will continue to review.

xiaoxmeng reviewed (commented):
@usurai thanks for the change. Left some comments and will continue to review.

usurai commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,67 +54,290 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+  types.reserve(numFields);
+  for (auto& field : joinCondition_->expr(0)->distinctFields()) {
+    const auto& name = field->field();
+    auto channel = leftType->getChildIdxIfExists(name);
```
`channel` is later used to get child index of `buildType` if `field` is not child of `probeType`.

xiaoxmeng commented on velox/exec/OperatorUtils.h:
```diff
@@ -136,4 +136,14 @@ folly::Range<vector_size_t*> initializeRowNumberMapping(
     memory::MemoryPool* pool);
 
+// Projects children of 'src' row vector to 'dest' row vector according to
+// 'projections' and 'mapping', the row number of projected children is
```
```
and 'mapping'. 'size' specifies number of projected rows in 'dest'.
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,124 @@ class NestedLoopJoinProbe : public Operator {
+
+  bool getBuildData(ContinueFuture* future);
+
+  // Get probe side rows count to process on next call to getCrossProduct()
```
```
// Calculates the number of probe rows to match with the build side vectors given the output batch size limit.
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,124 @@ class NestedLoopJoinProbe : public Operator {
+  // Get probe side rows count to process on next call to getCrossProduct()
+  vector_size_t getProbeCnt() const;
+
+  // Get cross product of next 'probeCnt' rows of input_ and all rows of
```
```
// Generates ...,  and all rows of build side vector at 
'buildIndex_' in 'buildData_'.
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,124 @@ class NestedLoopJoinProbe : public Operator {
+  // buildData_.value()[buildIndex].
+  // 'outputType' specifies the type of output.
+  // Projections from input_ and buildData_ to the output are specified by
+  // 'probeProjections' and 'buildProjections', respectively. Caller is
```
```
s/'buildProjections', respectively/'buildProjections' respectively/
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,124 @@ class NestedLoopJoinProbe : public Operator {
+  // Probe side state
   // Input row to process on next call to getOutput().
   vector_size_t probeRow_{0};
+  // probeCnt used in the last call to getCrossProduct(). Gets reset upon every
```
```
  // Records the number of probed rows in last getCrossProduct() call. It is reset upon each 'buildIndex_' update.
  vector_size_t numPrevProbedRows_{0};
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+  VELOX_CHECK(!hasProbedAllBuildData());
+
+  const auto buildSize = buildData_.value()[buildIndex_]->size();
+  const auto batchSize = probeCnt * buildSize;
```
```
s/batchSize/numOutputRows/
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,124 @@ class NestedLoopJoinProbe : public Operator {
+  // buildMatched_ accordingly.
+  RowVectorPtr doMatching(vector_size_t probeCnt);
+
+  // Update probeRow_ and buildIndex_ by advance probeRow_ by probeCnt.
```
```
// Updates ... by advancing ... Returns ...if 'buildIndex_' points to ...
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
-      buildIndex_ = 0;
-      input_.reset();
+  } while (!hasProbedAllBuildData() &&
+           !buildData_.value()[buildIndex_]->size());
```
```
&& buildData_.value()[buildIndex_]->empty()
```

usurai replied:
`BaseVector` does not have a `empty()` method, should we add one?

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+  DecodedVector decodedFilterResult;
+  decodedFilterResult.decode(*filterResult[0], filterInputRows_);
+
+  const vector_size_t resultSize = decodedFilterResult.size();
```
```
s/resultSize/maxOutputRows/
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+      initializeRowNumberMapping(buildOutMapping_, resultSize, pool());
+  auto probeIndices = probeIndices_->asMutable<vector_size_t>();
+  auto buildIndices = buildIndices_->asMutable<vector_size_t>();
+  int32_t outputSize{0};
```
```
s/outputSize/numOutputRows/
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+        decodedFilterResult.valueAt<bool>(i)) {
+      rawProbeOutMapping[outputSize] = probeIndices[i];
+      rawBuildOutMapping[outputSize] = buildIndices[i];
+      outputSize++;
```
nit: ++ numOutputRows

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+      outputSize++;
+    }
+  }
+  if (needsProbeMismatch(joinType_)) {
```
Shall we cache this flag above and set probeMatched_ in match check loop above? Otherwise we might set the same probe row many times which is the build vector size?

usurai replied:
My intention was to reduce branching, cuz in the end, we need to test / set some state for each row in decoded result.
Say we wanna check if certain probe row is matched in match check loop, we may take either of follow method:
1. As for each `buildSize` rows in decoded result, the probe row are the same. We have a bool `currentProbeRowMatched` for current probe row:
```
if needsProbeMismatch && !currentProbeRowMatched
    currentProbeRowMatched = true
```
then update probeMatched if probe row has `currentProbeRowMatched`

2. Have a variable `prevMatchedProbeRow`:
```
if needsProbeMismatch && currentMatchedProbeRow != prevMatchedProbeRow
    prevMatchedProbeRow = currentMatchedProbeRow
    probeMatched.setValid(currentMatchedProbeRow, true)
```

Method 1 tests `currentProbeRowMatched` `decodedResult.size()` times, method 2 tests `prevMatchedProbeRow` `decodedResult.size()` times, and maybe cache locality is also affected ?

xiaoxmeng replied:
Sorry, I bit lost of previous version. I am good with the current version. In general, we could optimize the code based on profiling data if the optimized code is more complicated to start with implementation wise.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+    }
+    probeMatched_.updateBounds();
+  }
+  if (needsBuildMismatch(joinType_)) {
```
ditto

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
 
-bool NestedLoopJoinProbe::isFinished() {
-  return buildSideEmpty_ || (noMoreInput_ && input_ == nullptr);
+  if (!outputSize) {
```
```
nit: if (numOutputRows == 0) {
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,135 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+  if (!outputSize) {
+    return nullptr;
+  }
+  auto output = BaseVector::create<RowVector>(outputType_, outputSize, pool());
```
```
return BaseVector::create<RowVector>
```

usurai replied:
Seems if no output would be generated, it's expected to return nullptr.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,67 +54,296 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+  return output;
+}
+
+void NestedLoopJoinProbe::finishMatchAndProbeMismatch() {
```
```
How about we call it finishProbeInput?

VELOX_CHECK_NOT_NULL(input_) before we reset it?
```

usurai replied:
Changed, thanks!

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,67 +54,296 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+  for (auto& peer : peers) {
+    auto op = peer->findOperator(planNodeId());
+    auto* probe = dynamic_cast<NestedLoopJoinProbe*>(op);
+    VELOX_CHECK(probe);
```
```
VELOX_CHECK_NOT_NULL(probe);
```

xiaoxmeng reviewed (commented):
Looks great % minors. Haven't looked at the test yet. Will continue. Thanks!

xiaoxmeng reviewed (commented):
Looks great % minors. Haven't looked at the test yet. Will continue. Thanks!

xiaoxmeng reviewed (commented):
Looks great % minors. Haven't looked at the test yet. Will continue. Thanks!

xiaoxmeng reviewed (commented):
Looks great % minors. Haven't looked at the test yet. Will continue. Thanks!

xiaoxmeng reviewed (commented):
Looks great % minors. Haven't looked at the test yet. Will continue. Thanks!

xiaoxmeng reviewed (commented):
Looks great % minors. Haven't looked at the test yet. Will continue. Thanks!

xiaoxmeng commented on velox/exec/tests/NestedLoopJoinTest.cpp:
```diff
@@ -231,29 +312,68 @@ TEST_F(NestedLoopJoinTest, zeroColumnBuild) {
+       {"u5", SMALLINT()},
+       {"u6", TINYINT()}});
+
+  auto leftVectors = makeBatches(60, 5, leftTypes, pool_.get());
```
Shall we test with all the join types supported? Thanks!

usurai replied:
Added, thanks!

usurai commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,67 +54,296 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+        return output;
+      }
+      case ProbeOperatorState::kRunning: {
+        if (input_ == nullptr) {
```
`driver` calls `getOutput()` before adding input to it, so it's expected for an operator to get `getOutput()` call without having input. For example, `HashProbe` has logic for handling nullptr input in `getOutput()`.

usurai commented:
Hi @xiaoxmeng, thanks for the detailed review! I've updated the PR according to comments, please take a look.

usurai commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,67 +54,296 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+    switch (state_) {
+      case ProbeOperatorState::kFinish:
+        return nullptr;
+      case ProbeOperatorState::kWaitForPeers: {
```
Agreed. I've merged kWaitForPeer into kRunning, and used a new method `processingBuildMismatch` to indicate the state.
```
  bool processingBuildMismatch() const {
    return state_ == ProbeOperatorState::kRunning && input_ == nullptr &&
        noMoreInput_;
  }
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +123,227 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+  if (isFinished()) {
     return nullptr;
   }
+  while (true) {
```
How about structure as?
```
RowVectorPtr output{nullptr};
while (output == nullptr) {
  break in some special case;
}
return output;
```

usurai replied:
Yeah it's better, updated.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +123,227 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+
+  lastProbe_ = true;
+  // From now on, buildIndex_ is used to indexing into buildMismatched_
+  buildIndex_ = 0;
```
Can we have a check here that buildIndex_  is already zero as reset by the last processed input?

usurai replied:
Indeed, replaced assignment with VELOX_CHECK_EQ.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +123,227 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+
+  buildVectors_ = std::move(buildData);
+  if (buildVectors_->empty()) {
+    // Build side is empty. Return empty set of rows and  terminate the pipeline
```
```
nit: Returns empty ... and terminates ...
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,138 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+  DecodedVector decodedFilterResult;
+  decodedFilterResult.decode(*filterResult[0], filterInputRows_);
+
+  const vector_size_t maxOutputRows_ = decodedFilterResult.size();
```
```
s/maxOutputRows_/maxOutputRows/
```

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,138 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+  if (numOutputRows == 0) {
+    return nullptr;
+  }
+  auto output =
```
```
simply as?
return BaseVector::create<RowVector>();
```

usurai replied:
Seems there is no `create` function that takes no parameter.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -97,60 +352,138 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
+      initializeRowNumberMapping(probeOutMapping_, maxOutputRows_, pool());
+  auto rawBuildOutMapping =
+      initializeRowNumberMapping(buildOutMapping_, maxOutputRows_, pool());
+  auto probeIndices = probeIndices_->asMutable<vector_size_t>();
```
```
auto* rawProbeIndices
auto* rawBuildIndices
```

xiaoxmeng commented on velox/exec/tests/NestedLoopJoinTest.cpp:
```diff
@@ -15,15 +15,54 @@
+    queryStr_ = "SELECT t0, u0 FROM t {} JOIN u ON t.t0 {} u.u0";
+  }
+
+  void setLeftType(RowTypePtr leftType) {
```
nit: can we use setProbeType for setLeftType and setBuildType for setRightType to be consistent with hash join test. Thanks!

usurai replied:
I've changed all the left/right naming to probe/build to be consistent with HashJoinTest, thanks!

xiaoxmeng reviewed (commented):
@usurai thanks for the update. LGTM except the probe state handling. Thanks!

xiaoxmeng reviewed (commented):
@usurai thanks for the update. LGTM except the probe state handling. Thanks!

usurai commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,130 @@ class NestedLoopJoinProbe : public Operator {
+        noMoreInput_;
+  }
+
+  void setState(ProbeOperatorState s) {
```
Yeah, I will do this as a followup once we've clarified the state transition.

usurai commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,48 +54,65 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+        return BlockingReason::kNotBlocked;
+      }
+      if (future_.valid()) {
+        *future = std::move(future_);
```
Agreed. I removed `kWaitForPeers` state but it turns out that it applies to operators that:
1. Perform right or full joins.
2. Have completed processing the probe input.
3. Are not the last operator to finish processing the probe input.

In the new version, aforementioned operators enter `kWaitForPeers` state at `beginBuildMismatch()`, and turn into `kFinish` state at `isBlocked`. Meanwhile, the last probe operator goes from `kRunning` directly into `kFinish`.

usurai commented on velox/exec/tests/OperatorUtilsTest.cpp:
```diff
@@ -297,3 +299,66 @@ TEST_F(OperatorUtilsTest, initializeRowNumberMapping) {
+      {"string_val", VARCHAR()},
+  });
+  VectorFuzzer fuzzer({}, pool());
+  auto srcVector{fuzzer.fuzzFlat(srcRowType, srcVectorSize)};
```
I didn't know about this method, thanks!

usurai commented on velox/exec/NestedLoopJoinProbe.h:
```diff
@@ -31,30 +32,130 @@ class NestedLoopJoinProbe : public Operator {
+
+  // Wraps rows of 'data' that are not selected in 'matched' and projects
+  // to the children of output according to 'projections'. Creates null vector
+  // as output's children according to 'nullProjections'. 'mapping' is used as
```
Thanks for the refinement!

usurai commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +123,227 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
     return nullptr;
   }
+  while (true) {
+    VELOX_CHECK_EQ(state_, ProbeOperatorState::kRunning);
```
Indeed, removed.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +122,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+      break;
+    }
+
+    if (input_ == nullptr) {
```
I am curious when will we run into null input here?

usurai replied:
On `driver`'s first `getOutput()` call to the operator:
https://github.com/facebookincubator/velox/blob/333b870cf2d4668c4b4dfff29c1450316081b65d/velox/exec/Driver.cpp#L399

When processing a pipeline, `driver` does the following to operators, in reverse order(from sink to source):
```
// this is simplified
if !currentOperator.isBlocked()
    if nextOperator.needsInput()
        output = currentOperator.getOutput()
        if output
            nextOperator.addInput(output)
            currentOperator = nextOperator
        else
            check if currentOperator is blocked or finished
            if not, currentOperator = prevOperator
```

As `driver` always tries to find the last operator that can emit output, `getOutput` is called to see if it can output something before input is added to the operator, the `if` is for this case.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +122,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+      break;
+    }
+
+    vector_size_t probeCnt = getNumProbeRows();
```
nit: const vector_size_t probeCnt =

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +122,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+    // matching of input_ and buildData_ has finished. For left/full joins,
+    // the next step is to emit output for mismatched probe side rows.
+    if (hasProbedAllBuildData()) {
+      output = needsProbeMismatch(joinType_) ? getMismatchedOutput(
```
It seems that needsProbeMismatch is always true in this case? Shall we add CHECK instead?

usurai replied:
> It seems that needsProbeMismatch is always true in this case?

The only case it's false I can think of is empty build data.

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +122,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+
+  std::vector<ContinuePromise> promises;
+  std::vector<std::shared_ptr<Driver>> peers;
+  if (!operatorCtx_->task()->allPeersFinished(
```
It looks like we don't need kWaitForPeers state as we don't need to wait here. We can simply pass null future as HashProbe::noMoreInputInternal() does. Then if it is not the last prober, then we could simply set state to finish and return.

usurai replied:
That' true.
At first, I thought if you need to grab some state from an operator, it needs to be unfinished. Turns out it don't need to be...

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -28,48 +54,64 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
+      return BlockingReason::kNotBlocked;
+    }
+    case ProbeOperatorState::kWaitForPeers: {
+      if (future_.valid()) {
```
We probably doesn't need future_ see comment below. thanks!

usurai replied:
Removed, thanks!

xiaoxmeng reviewed (approved):
@usurai thanks for the iterations!

xiaoxmeng reviewed (approved):
@usurai thanks for the iterations!

usurai commented:
@xiaoxmeng Thanks for review!

xiaoxmeng commented:
> @xiaoxmeng Thanks for review!

@usurai can you take a look the last few comments? Then I can go ahead to merge this diff. Thanks!

xiaoxmeng commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +113,224 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+  std::vector<std::shared_ptr<Driver>> peers;
+  ContinueFuture future{ContinueFuture::makeEmpty()};
+  if (!operatorCtx_->task()->allPeersFinished(
+          planNodeId(), operatorCtx_->driver(), &future, promises, peers)) {
```
I think we can pass future as null as we don't need to wait. And also check promises are empty on the last prober? thanks!

usurai replied:
Done, thanks!

xiaoxmeng commented:
> @xiaoxmeng Sorry I somehow overlooked these. Have updated.

@usurai thanks for the quick update!

mbasmanova commented:
@usurai Thank you for working on this. The PR description seems outdated and no longer matches the PR content. Any chance you could update it?

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -224,12 +224,12 @@ and emitting results.
 
 NestedLoopJoinNode
-~~~~~~~~~~~~~~~~~~
+~~~~~~~~~~~~~
```
Unintended change? CC: @majetideepak

usurai replied:
Sorry, I may have done it incorrectly during the process of rebasing.

mbasmanova replied:
No problem. Thank you for fixing.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -238,6 +238,10 @@ input, the output of the cross join will contain N * M rows.
+   * - joinType
+     - Join type: inner, left, right, full.
+   * - joinCondition
+     - SQL expression as the join condition. Can use columns from both probe and build sides of the join.
```
Using "SQL" here might be confusing as Velox doesn't support SQL text.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -241,7 +241,7 @@ and emitting results. Nested loop join supports non-equality join.
      - Join type: inner, left, right, full.
    * - joinCondition
-     - SQL expression as the join condition. Can use columns from both probe and build sides of the join.
+     - Expression used as the join condition, may reference cololumns from both inputs.
```
Nice. 

typo: cololumns

mbasmanova commented on velox/exec/NestedLoopJoinProbe.cpp:
```diff
@@ -80,15 +113,221 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+}
+
+RowVectorPtr NestedLoopJoinProbe::getMismatchedOutput(
+    RowVectorPtr data,
```
use const &; please, go over the PR and make sure to pass shared pointers by reference.

usurai replied:
Sure, I've read through the changes and updated with the fix.

mbasmanova reviewed (commented):
@usurai Some comments.

usurai commented on velox/exec/OperatorUtils.h:
```diff
@@ -136,4 +136,14 @@ folly::Range<vector_size_t*> initializeRowNumberMapping(
     vector_size_t size,
     memory::MemoryPool* pool);
 
+// Projects children of 'src' row vector to 'dest' row vector according to
```
It means transforming the children of `src` specified in `projections`' input channels to the children of `dest` specified in `projections`' output channels, by:
1. If `mapping` is nullptr, the child vector stays identical
2. If `mapping` is not null, the dest child vector is a wrapped dictionary vector of the src child vector according to `mapping`.

usurai replied:
@mbasmanova, I created a PR that updates `fillOutput` to make use of `projectChildren`, please take a look. #5127
